lisa/src/main/java/it/unive/lisa/analysis/heap/PointBasedHeap.java

```diff
@@ -0,0 +1,247 @@
+ * their allocation sites, namely the position of the code wehre heap locations
+ * are generated. All heap locations that are generated at the same allocation
+ * sites are abstracted into a single unique identifier.
+ * 
+ * @author <a href="mailto:vincenzo.arceri@unive.it">Vincenzo Arceri</a>
+ */
+public class PointBasedHeap extends BaseLattice<PointBasedHeap> implements HeapDomain<PointBasedHeap> {
+
+	private static final PointBasedHeap TOP = new PointBasedHeap();
+
+	private static final PointBasedHeap BOTTOM = new PointBasedHeap();
+
+	private final Collection<ValueExpression> rewritten;
+
+	private final Map<Identifier, Collection<ValueExpression>> allocationSites;
+
+	/**
+	 * Builds a new instance of PointBasedHeap, with an unique rewritten
+	 * expression {@link Skip}.
+	 */
+	public PointBasedHeap() {
+		this(new Skip());
+	}
+
+	private PointBasedHeap(ValueExpression rewritten) {
+		this(Collections.singleton(rewritten), new HashMap<>());
+	}
+
+	private PointBasedHeap(Collection<ValueExpression> rewritten,
+			Map<Identifier, Collection<ValueExpression>> allocationSites) {
+		this.rewritten = rewritten;
+		this.allocationSites = allocationSites;
+	}
+
+	@Override
+	public PointBasedHeap smallStepSemantics(SymbolicExpression expression, ProgramPoint pp) throws SemanticException {
+
+		if (expression instanceof AccessChild) {
+			PointBasedHeap container = smallStepSemantics((((AccessChild) expression).getContainer()), pp);
+			return new PointBasedHeap(container.getRewrittenExpressions(), allocationSites);
+		}
+
+		if (expression instanceof HeapAllocation) {
+			HeapIdentifier id = new HeapIdentifier(expression.getTypes(), pp.toString(), true);
+			return new PointBasedHeap(Collections.singleton(id), allocationSites);
+		}
+
+		if (expression instanceof HeapReference)
+			// TODO: not sure about this (need to check)
+			return this;
+
+		if (expression instanceof UnaryExpression) {
+			UnaryExpression unary = (UnaryExpression) expression;
+			PointBasedHeap sem = smallStepSemantics(unary.getExpression(), pp);
+			PointBasedHeap result = bottom();
+			for (ValueExpression expr : sem.getRewrittenExpressions())
+				result = result.lub(mk(sem, new UnaryExpression(expression.getTypes(), expr, unary.getOperator())));
+			return result;
+		}
+
+		if (expression instanceof BinaryExpression) {
+			BinaryExpression binary = (BinaryExpression) expression;
+			PointBasedHeap sem1 = smallStepSemantics(binary.getLeft(), pp);
+			PointBasedHeap sem2 = sem1.smallStepSemantics(binary.getRight(), pp);
+			PointBasedHeap result = bottom();
+			for (ValueExpression expr1 : sem1.getRewrittenExpressions())
+				for (ValueExpression expr2 : sem2.getRewrittenExpressions())
+					result = result.lub(
+							mk(sem2, new BinaryExpression(expression.getTypes(), expr1, expr2, binary.getOperator())));
+			return result;
+		}
+
+		if (expression instanceof ValueIdentifier)
+			if (allocationSites.containsKey(expression))
```

Comment:
Why do we rewrite `ValueIdentifier`s? This should happen only with `HeapExpression`s